### ISSUE TITLE
PAL: safecrt_output_l.c contains assertion which errors/fails on FreeBSD

### ISSUE THREAD

**Issue by josteink** [dotnet/coreclr#679]
This causes the build to break.

Output from build:

````
[  7%] Building C object src/pal/src/CMakeFiles/CoreClrPal.dir/safecrt/safecrt_output_l.c.o
/home/josteink/build/cc/src/pal/src/safecrt/safecrt_output_l.c:136:5: error: 'sanity_check_dummy' declared as an array \
with a negative size
    CASSERT(sizeof(long double) == sizeof(double));
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/josteink/build/cc/src/pal/src/safecrt/mbusafecrt_internal.h:48:50: note: expanded from macro 'CASSERT'
#define CASSERT(p) extern int sanity_check_dummy[1+((!(p))*(-2))];
                                                 ^~~~~~~~~~~~~~~
1 error generated.
````

The following test-program confirms the error with the following rather obvious results:

````
➜  ~  cat test.c
#include <stdlib.h>
#include <stdio.h>

void main(void) {
  printf("%d\n", sizeof(long double));
  printf("%d\n", sizeof(double));
}
➜  ~  clang test.c
...
➜  ~  ./a.out
16
8
````

This is part of a more [intricate ifdef-section](https://github.com/dotnet/coreclr/blob/master/src/pal/src/safecrt/safecrt_output_l.c#L127-L137) of `safecrt_output_l.c`:

````c
#if (defined(__DARWIN_LONG_DOUBLE_IS_DOUBLE) && !__DARWIN_LONG_DOUBLE_IS_DOUBLE) || defined(__LINUX__) || defined(__APPLE__)
    #ifdef MBUSAFECRT_SUPPORTS_LONGDOUBLE   // we do not have _cldcvt function to convert long double.
        #define LONGDOUBLE_IS_DOUBLE 0  /* 1 means long double is same as double */
        CASSERT(sizeof(long double) != sizeof(double));
    #else
        #define LONGDOUBLE_IS_DOUBLE 1
    #endif
#else
    #define LONGDOUBLE_IS_DOUBLE 1  /* 1 means long double is same as double */
    CASSERT(sizeof(long double) == sizeof(double));
#endif
````

From what I can tell, the correct code-path for FreeBSD would involve the section defining LONGDOUBLE_IS_DOUBLE to 0, and the current code pressumes this to be checked/covered by `MBUSAFECRT_SUPPORTS_LONGDOUBLE`.

I cannot however find any references to `MBUSAFECRT_SUPPORTS_LONGDOUBLE` or obviously related material anywhere either, so I'm not sure where to proceed.

Any advice would be appreciated.

**Comment by janvorli** [comment]
@josteink Let me take a look.

**Comment by josteink** [comment]
For the time being I've short-circutted the checks and enforced a FreeBSD-specific value.

Basically this commit: https://github.com/josteink/coreclr/commit/5881603d08ef384e45db86dcd5a7f4ab35ee0582

To me though, that doesn't look like a proper solution, or at least not one I'm happy pushing upstream until I've had it verified by others.

**Comment by janvorli** [comment]
I've taken a look and I can see that PAL doesn't support printing long double values when the long double is different from double. Although there is an internal `FL_LONGDOUBLE` flag that the code uses when the  `LONGDOUBLE_IS_DOUBLE` is zero, that flag is never set, because it doesn't detect `L` prefix for the floating point format specifiers to indicate a long double parameter.
I guess that's due to the fact that in VC++, long double and double are the same size (8 bytes) and the formatting functions were ported from windows versions.
Moreover, besides the `safecrt_output_l.c`, the `LONGDOUBLE_IS_DOUBLE` is also defined in the `output.inl` file. In that place, right after the same complex `#if` that you have found in the `safecrt_output_l.c`, the `LONGDOUBLE_IS_DOUBLE` is defined again, this time uncoditionally to 1.
I think that we should base the `LONGDOUBLE_IS_DOUBLE` on the sizeof(double) to sizeof(long double) comparison instead of the platform and other stuff in that complex `#if`. That means we should add the detection to the `configure.cmake`
I have also verified the same way you did that even on Linux the long double is 16 bytes long, so it is different from the double which is 8 bytes long. 
@jkotas, do you have any insight on the history of the way the detection was made and what the `MBUSAFECRT_SUPPORTS_LONGDOUBLE` really means?

**Comment by jkotas** [comment]
I do not have the history for this one. Based on the name, this define was likely meant to support Office for Mac - MBU likely stands for Macintosh Business Unit.

> I think that we should base the  LONGDOUBLE_IS_DOUBLE  on the sizeof(double) to sizeof(long double)
 
Agree.

**Comment by josteink** [comment]
Just a few questions:

* Should we keep the definition local to the file, or should it be promoted to `configure.cmake`?
* Who makes it happen? Do you want me to prepare a PR for this, or do you want to handle it yourself?

Apart from that, it all sounds good to me.

**Comment by janvorli** [comment]
@josteink I will do it. It seems to me that putting it to the configure.cmake is the right thing to do.

**Comment by janvorli** [comment]
Fixed by #715

**Comment by josteink** [comment]
Just confirming that PR #715 fixes this error. I must say: That seems like quite a bit more work than I thought would be required to resolve this one

Thanks for the effort and thanks for the help! :)
